presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkTaskExecution.java

```diff
@@ -0,0 +1,471 @@
+import com.facebook.presto.operator.DriverFactory;
+import com.facebook.presto.operator.DriverStats;
+import com.facebook.presto.operator.PipelineContext;
+import com.facebook.presto.operator.TaskContext;
+import com.facebook.presto.spi.plan.PlanNodeId;
+import com.facebook.presto.sql.planner.LocalExecutionPlanner.LocalExecutionPlan;
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableMap;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import com.google.common.util.concurrent.ListenableFuture;
+import io.airlift.units.Duration;
+
+import javax.annotation.Nullable;
+import javax.annotation.concurrent.GuardedBy;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.OptionalInt;
+import java.util.Set;
+import java.util.concurrent.Executor;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.facebook.presto.SystemSessionProperties.getInitialSplitsPerNode;
+import static com.facebook.presto.SystemSessionProperties.getMaxDriversPerTask;
+import static com.facebook.presto.SystemSessionProperties.getSplitConcurrencyAdjustmentInterval;
+import static com.facebook.presto.operator.PipelineExecutionStrategy.UNGROUPED_EXECUTION;
+import static com.google.common.base.MoreObjects.toStringHelper;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.base.Verify.verify;
+import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.stream.Collectors.collectingAndThen;
+import static java.util.stream.Collectors.groupingBy;
+import static java.util.stream.Collectors.mapping;
+
+/**
+ * The PrestoSparkTaskExecution is a simplified version of SqlTaskExecution.
+ * It doesn't support grouped execution that is not needed on Presto on Spark.
+ * Unlike the SqlTaskExecution the PrestoSparkTaskExecution does not require
+ * the output buffer to be drained to mark the task as finished. As long as
+ * all driver as finished the task execution is marked as finished. That allows to
+ * have more control over the output Iterator lifecycle in the PrestoSparkTaskExecutor
+ */
+public class PrestoSparkTaskExecution
+{
+    private final TaskId taskId;
+    private final TaskStateMachine taskStateMachine;
+    private final TaskContext taskContext;
+
+    private final TaskHandle taskHandle;
+    private final TaskExecutor taskExecutor;
+
+    private final Executor notificationExecutor;
+
+    private final SplitMonitor splitMonitor;
+
+    private final List<PlanNodeId> schedulingOrder;
+    private final Map<PlanNodeId, DriverSplitRunnerFactory> driverRunnerFactoriesWithSplitLifeCycle;
```

Comment:
FWIW, before grouped execution is introduced in 4e55aadca50e2b6f904fcfa444540e90c02e3383, they are called `partitionedDriverRunnerFactories` and `unpartitionedDriverRunnerFactories` :) . We can also call them `tableScanDriverRunnerFactories` vs. `exchangeDriverRunenrFactories`? :)

Reply:
`tableScanDriverRunnerFactories` sounds good. However `exchangeDriverRunenrFactories` is not necessarily accurate, as for example aggregation will have task lifecycle runners.

Reply:
>  as for example aggregation will have task lifecycle runners.

don't the drivers with task lifecycle runners read from *local* exchange ?

Reply:
Actually you are right. In order for pipeline to be task wide, the pipeline must either read from a local or from a remove exchange. Let me do the rename.

Reply:
After some thinking I realize that `exchangeDriverRunenrFactories` and `tableScanDriverRunnerFactories` might sound somehow misleading. The name may imply that the `DriverRunners` will be running only table scans or only exchanges, when in practice they run some other operators as part of the same pipeline.

Reply:
@arhimondr : > The name may imply that the DriverRunners will be running only table scans or only exchanges, when in practice they run some other operators as part of the same pipeline.

That's a fair point. I don't think there is good name: 

- The old "partitioned" / "unpartitioned" name is cryptic to understand. But it's also quite universal and some engineers (like me) already read them as "table scan driver" vs. "exchange driver"

- `driverRunnerFactoriesWithSplitLifeCycle` and `driverRunnerFactoriesWithTaskLifeCycle` is correct, but also not easy to understand.. in fact I did a manual map to "partitioned driver runner" vs. "unpartitioned driver runner"

- `tableScanDriverRunnerFactories` / `exchangeDriverRunenrFactories` : It's easiest for me to understand, but as you point out, the name is misleading... `tableScanSourcedDriverRunnerFactories` / `exchangeSourcedDriverRunenrFactories` is more accurate but also too verbose... 😕 

So... let's keep what you named there... 😃